### PR TITLE
Transfer from ZMQ to WebSockets

### PR DESCRIPTION
**Pull request by trushton**
@kevgo @hugobho @martinjaime @helenmedrano 
This PR is dependent on the Exocom-mock-js PR, as it uses that repo for testing. 
@hugobho if you can make sure this all looks good to match up with your work on ExoCom-dev that would be great. 
All comments and ideas for improvement are appreciated.

### CODE REVIEW

File: src/websocket-connector/websocket-connector.ls [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  'events' : {EventEmitter}
+  'lodash.isempty' : is-empty
+  'node-uuid' : uuid
+  'ws' : WebSocket
+}
+debug = require('debug')('exorelay:websocket-listener')
+
+
+# The WebSocket endpoint into which the Exosphere environment can
+# PUSH new messages.
```
Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: Messages can be sent and received from this endpoint now, not just pushed.

File: src/websocket-connector/websocket-connector.ls [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#
+# Emits these events:
+# - online
+# - offline
+# - error
+class WebSocketConnector extends EventEmitter
+
+  ({@exocom-host, @service-name, @exocom-port} = {}) ->
+
+    @exocom-port = +@exocom-port
```
Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: I think you should just be able to do `({@exocom-host, @service-name, +@exocom-port} = {}) ->` in the constructor.

File: src/websocket-connector/websocket-connector.ls [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        @emit 'online'
+      ..on 'message', (message) ~>
+        @_on-message message
+      ..on 'error' (error) ~>
+        | error.errno is 'EADDRINUSE'  =>  @emit 'error', "port #{@exocom-port} is already in use"
+        | otherwise                    =>  @emit 'error', error
+
+
+  _on-message: (data) ~>
+    request-data = data |> (.to-string!) |> JSON.parse |> @_parse-request
```
Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: Does this work without the `(.to-string!)` method? I think the `ws` module sends/receives in string format by default.

File: src/websocket-connector/websocket-connector.ls [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+# Emits these events:
+# - online
+# - offline
+# - error
+class WebSocketConnector extends EventEmitter
+
+  ({@exocom-host, @service-name, @exocom-port} = {}) ->
+
+    @exocom-port = +@exocom-port
+    @listen!
```
Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: I think we should remove this from the constructor and call it from `exorelay.ls` i.e. when we do `delegate \close \listen from: @, to: @websocket-connector`.

File: src/exorelay.ls [hunk 5]
```diff
@@ -19,48 +19,42 @@ class ExoRelay extends EventEmitter
...
 
 
-    delegate \closePort, from: @, to: @message-sender
-    delegate \close \listen \port from: @, to: @zmq-listener
+    delegate \close \listen from: @, to: @websocket-connector
     delegate \hasHandler \registerHandler \registerHandlers from: @, to: @message-handler
-    delegate-event 'error', from: [@zmq-listener, @message-handler, @message-sender], to: @
-    delegate-event 'status', 'offline', from: @zmq-listener, to: @
+    delegate-event 'error', from: [@websocket-connector, @message-handler], to: @
+    delegate-event 'status', 'offline', from: @websocket-connector, to: @
```
Comment (hugobho) [inline src/exorelay.ls]: I don't think we actually emit 'status' anymore.

File: src/exorelay.ls [hunk 6]
```diff
@@ -2,9 +2,9 @@ require! {
   'events' : {EventEmitter}
   'ip'
```
Comment (hugobho) [inline src/exorelay.ls]: Let's get rid of this in a separate PR.

File: features/listen.feature [hunk 7]
```diff
@@ -15,13 +15,5 @@ Feature: Listening
...
-    When I take it online at port 4001
-    Then it is online at port 4001
-    And it emits the 'online' event with payload 4001
-
-
-  # ERROR HANDLING
-
-  Scenario: providing a non-number as the port
-    When I try to take it online at port "zonk"
-    Then ExoRelay emits an "error" event with the error "Non-numerical port provided to ExoRelay#listen"
```
Comment (hugobho) [inline features/listen.feature]: Why did you get rid of all this? It seems like we can still set up a ExoRelay at a given port. Just that if we want it to talk to ExoCom then it better be the ExoCom port. Although I do agree that ExoRelay doesn't really "listen" at a specific port so much as it binds to ExoCom at a port and both sends/receives on that port.

File: features/listen.feature [hunk 8]
```diff
@@ -15,13 +15,5 @@ Feature: Listening
 
 
   Scenario: Setting up at the given port
-    When I take it online at port 4001
-    Then it is online at port 4001
-    And it emits the 'online' event with payload 4001
```
Comment (hugobho) [inline features/listen.feature]: Should it not still emit an 'online' event after you take it online?

File: features/receiving-messages.feature [hunk 9]
```diff
@@ -11,7 +11,7 @@ Feature: Receiving messages
 
   Background:
     Given ExoCom runs at port 4100
-    And an ExoRelay instance listening on port 4000
+    And an ExoRelay instance listening
```
Comment (hugobho) [inline features/receiving-messages.feature]: `And an ExoRelay instance`

File: features/steps/port-steps.ls [hunk 10]
```diff
@@ -13,26 +13,15 @@ module.exports = ->
       ..listen!
 
 
-  @When /^I take it online at port (\d+)$/, (port, done) ->
+  @When /^I take it online$/, (done) ->
     @exo-relay
       ..on 'online', ~>
-        wait-until (~> @exocom.received-messages.length), 10, ~>
-          @exocom.reset!
+        wait-until (~> @exocom.service-sockets['test-service']), 10, ~>
```
Comment (hugobho) [inline features/steps/port-steps.ls]: 👍

File: features/listen.feature (hunk 7 above)
Comment (trushton) [inline features/listen.feature]: Seeing as ExoRelay no longer needs its own port to listen on, we just connect to the ExoCom server, this scenario becomes irrelevant. There is no ExoRelay port to provide a non-number to.

File: features/listen.feature (hunk 8 above)
Comment (trushton) [inline features/listen.feature]: This step simply checks that the payload, @online-port, is equal to the value provided. Since we no longer have and @online-port number, this step isn't doing anything.

File: src/websocket-connector/websocket-connector.ls (hunk 4 above)
Comment (trushton) [inline src/websocket-connector/websocket-connector.ls]: My understanding was that as soon as a service boots up, it would spin up its ExoRelay instance and dial in automatically to the ExoCom WebSocketServer. Are there any other times that we would want to tell ExoRelay to `listen` other than right at start-up?

File: src/websocket-connector/websocket-connector.ls (hunk 2 above)
Comment (trushton) [inline src/websocket-connector/websocket-connector.ls]: Running it like that gives `TypeError: Function.prototype.toString is not generic`

File: src/websocket-connector/websocket-connector.ls (hunk 4 above)
Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: Would there ever be a case where we'd want to `close` ExoRelay and then `listen` again some time later? Extracting a `listen` function out would allow us to do that without creating a whole new ExoRelay object.

File: features/listen.feature (hunk 7 above)
Comment (hugobho) [inline features/listen.feature]: But a port number is still passed into ExoRelay via ExoService, where it's set using environment variables right? Even though it's supposed to be the ExoCom port the user could still pass non-sense to ExoRelay.

You're still verifying that it's a number in the `listen` function of `websocket-connector.ls`.

File: features/listen.feature (hunk 8 above)
Comment (hugobho) [inline features/listen.feature]: This will apply again now, no? See discussion below on `features/listen.feature`.

File: features/receiving-messages.feature (hunk 9 above)
Comment (hugobho) [inline features/receiving-messages.feature]: This comment was unaddressed

File: features/register-service.feature [hunk 11]
```diff
@@ -12,15 +12,13 @@ Feature: Register a service with ExoCom
     Given ExoCom runs at port 4100
 
   Scenario: registering a service with ExoCom
-    When an ExoRelay instance running inside the "test-service" service comes online at port 4000
-    Then ExoRelay makes the ZMQ request:
+    When an ExoRelay instance running inside the "test-service" service comes online
+    Then ExoRelay makes the WebSocket request:
```
Comment (hugobho) [inline features/register-service.feature]: In the Rules section of this `register-service.feature` file we still say that we're coming online via a ZMQ message. We should change that to WebSockets

**Comment by hugobho** [comment]
Looks good, a couple more changes to make. Most notably probably extracting a `listen` method, depending on what you think about that discussion. And also still checking that ExoRelay received a proper port and emitting an event for that. Also some comments about small style changes were missed.

File: features/register-service.feature (hunk 11 above)
Comment (trushton) [inline features/register-service.feature]: That change has already been made

File: features/listen.feature (hunk 7 above)
Comment (trushton) [inline features/listen.feature]: Thats true about ExoService. So that `EXORELAY_PORT` from ExoService really ought to be removed in another PR so that it is not even passed into ExoRelay. 

There could still be use for this test I suppose, we can check that the ExoCom port is valid as opposed to the ExoRelay port that was being tested before.

File: src/websocket-connector/websocket-connector.ls (hunk 4 above)
Comment (trushton) [inline src/websocket-connector/websocket-connector.ls]: I may be misinterpreting your comment, but a `listen` function is already extracted out. We could still call `exorelay.listen!` and thanks to the delegate it would exorelay would call the `listen` function inside of its `@websocket-connector`. The same goes for the `close` function. As for when we would want to close and then listen again later? I'm not entirely sure of that use case.

Comment (hugobho) [inline src/websocket-connector/websocket-connector.ls]: Okay that makes sense. Yeah I guess I can't think of any case where we wouldn't want to listen right after constructing ExoRelay. Let's stick with the way things are for now.

File: features/listen.feature (hunk 7 above)
Comment (hugobho) [inline features/listen.feature]: Yeah that's what I meant. There's a line in `websocket-connector.listen` that checks if the port is a number. It used verify for the pull socket port but now it can be used to verify the ExoCom port.

Users can still boot up services independent of `exo-run` by using Exoservice (executing `exo-js` in service directory). In those cases the user has to provide the ExoCom port as an environment variable so it's useful to have this test still.

Comment (hugobho) [inline features/listen.feature]: Although I'm not sure when an application would want to use an independent Exoservice that wasn't listed in its `application.yml` file. On second thought, let's just go with the assumption that services are started via `exo-run` or `exo-deploy` only and hence always receive a valid port number.

File: features/register-service.feature [hunk 12]
```diff
@@ -5,22 +5,20 @@ Feature: Register a service with ExoCom
   So that services can scale up and down automatically.
 
   Rules:
-  - exoRelay registers itself with ExoCom when coming online via the ZMQ message "exocom.register-service"
+  - exoRelay registers itself with ExoCom when coming online via the WebSocket message "exocom.register-service"
```
Comment (kevgo) [inline features/register-service.feature]: Can this be merged with features/listen.feature?

```cucumber
Given ExoCom runs at port 4100
When taking an ExoRelay instance online 
Then it makes this WebSocket request to http://localhost:4100:
  """
  ...
  """

File: features/steps/exocomm-address-step.ls [hunk 13]
```diff
@@ -28,14 +29,15 @@ module.exports = ->
       @error = e.message
 
 
-  @When /^trying to create an ExoRelay without providing the ExoCom host$/ ->
+  @When /^trying to create an ExoRelay without providing the ExoCom host$/, ->
```
Comment (kevgo) [inline features/steps/exocomm-address-step.ls]: LiveScript is smart enough to add commas where it is obvious one is needed, like here.

File: features/steps/instance-steps.ls [hunk 14]
```diff
@@ -13,59 +13,43 @@ require! {
...
-    @exocom.register-service name: 'test-service', port: port
+  @Given /^an ExoRelay instance$/, (done) ->
     @exo-relay = new ExoRelay {exocom-host: 'localhost', @exocom-port, service-name: 'test-service'}
-      ..on 'online', (@online-port) ~>
+      ..on 'online', ~>
         wait-until (~> @exocom.received-messages.length), 10, ~>
           @exocom.reset! if @exocom.received-messages |> any (.name is 'exocom.register-service')
           done!
       ..on 'error', (@error) ~>
-      ..listen port
```
Comment (kevgo) [inline features/steps/instance-steps.ls]: If you don't call `listen` here, the exorelay instance never goes online, and then it never calls `done` here.

File: features/steps/instance-steps.ls [hunk 15]
```diff
@@ -13,59 +13,43 @@ require! {
...
-
-
-  @Given /^an ExoRelay instance$/, ->
-    @exo-relay = new ExoRelay {exocom-host: 'localhost', @exocom-port, service-name: 'test-service'}
-      ..on 'online', (@online-port) ~>
 
 
-  @Given /^an ExoRelay instance listening on port (\d+)$/, (port, done) ->
-    @exocom.register-service name: 'test-service', port: port
+  @Given /^a new ExoRelay instance$/, ->
```
Comment (kevgo) [inline features/steps/instance-steps.ls]: What is the difference between `an ExoRelay instance` and `a new ExoRelay instance`?

File: features/steps/port-steps.ls (hunk 10 above)
Comment (kevgo) [inline features/steps/port-steps.ls]: How does it know to use `test-service` here? The MockExoCom should still record all received messages, and we should still use them here.

File: features/listen.feature (hunk 7 above)
Comment (kevgo) [inline features/listen.feature]: ExoCom host and port are still provided to ExoRelay, so they are part of its API, and it should be documented here that they are handled correctly. This means describing a scenario when correct values are given, a scenario when incorrect values are given (wrong port number).

File: src/exorelay.ls [hunk 16]
```diff
@@ -1,10 +1,9 @@
 require! {
   'events' : {EventEmitter}
-  'ip'
   './message-handler/message-manager' : HandlerManager
-  './message-sender/message-sender' : MessageSender
   'rails-delegate' : {delegate, delegate-event}
-  './zmq-listener/zmq-listener' : ZmqListener
+  './websocket-connector/websocket-connector' : WebSocketConnector
+  'ws' : WebSocket
```
Comment (kevgo) [inline src/exorelay.ls]: You don't use `WebSocket` in this file, remove this line

File: src/exorelay.ls [hunk 17]
```diff
@@ -19,48 +18,43 @@ class ExoRelay extends EventEmitter
...
-    @zmq-listener = new ZmqListener!
+    # send and receives messages from Exosphere
+    @websocket-connector = new WebSocketConnector config
       ..on 'message', @_on-incoming-message
       ..on 'online', @_send-routing-config
 
 
-    delegate \closePort, from: @, to: @message-sender
-    delegate \close \listen \port from: @, to: @zmq-listener
+    # TODO: Decide whether or not the listen event is needed, or if it will be ran automatically on the start-up of the @websocket-connector
```
Comment (kevgo) [inline src/exorelay.ls]: Constructors should not do any work. So making a network request should be done in a separate method like `listen` or `connect`.

File: src/websocket-connector/websocket-connector-spec.ls [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+
+  describe 'reply-method-for', (...) ->
+
+    before-each ->
+      @websocket-connector.send = sinon.stub!
+      @reply-method = @websocket-connector.reply-method-for '123'
+
+    it 'returns a function', ->
+      expect(typeof @reply-method).to.equal 'function'
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector-spec.ls]: expect(@reply-method).to.be.a 'function'

File: src/websocket-connector/websocket-connector-spec.ls [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+    it 'calls @send', ->
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector-spec.ls]: I would combine all 3 tests into 

```ls
it 'returns a function that calls @send prebound to the response id'
```

Yes, its not ideal, but I don't see another way to break this up and be readable.

File: src/websocket-connector/websocket-connector.ls [hunk 20]
```diff
@@ -0,0 +1,101 @@
+require! {
+  'events' : {EventEmitter}
+  'lodash.isempty' : is-empty
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Please don't mix lodash and prelude-ls. Use the latter. Also, this doesn't seem to be used, so remove.

File: src/websocket-connector/websocket-connector.ls [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+# send messages and receive messages from.
+#
+# Emits these events:
+# - online
+# - offline
+# - error
+class WebSocketConnector extends EventEmitter
+
+  ({@exocom-host, @service-name, @exocom-port} = {}) ->
+
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Remove this empty line

File: src/websocket-connector/websocket-connector.ls (hunk 4 above)
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: A time when we just want to boot up an ExoRelay without it automatically trying to connect to Exocom is in tests. Maybe we want to check another method (let's call it "foo") on the ExoRelay instance. If ExoRelay auto-connects, our test for "foo" need to create an ExoCom instance, which is completely unrelated to "foo". Now our tests are cumbersome and full of boilerplate. That's why constructors should not do any real work. They should only set up the object graph, and graciously accept `null` for any of their dependencies.

File: src/websocket-connector/websocket-connector.ls [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+    @listen!
+
+    # Contains the id of the most recently sent request (for testing)
+    @last-sent-id = null
+
+
+  # Closes the port that ExoRelay is communicating on
+  close: ->
+    return unless @socket
+    debug "no longer istening at port #{@exocom-port}"
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Typo in `istening`

File: src/websocket-connector/websocket-connector.ls [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+    @socket.close!
+    @emit 'offline'
+
+
+  # Returns a method that sends a reply to the message with the given request
+  reply-method-for: (id) ->
+    | !id  =>  return @emit 'error', new Error 'WebSocketConnector.replyMethodFor needs an id'
+
+    (message-name, payload) ~>
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Use LiveScript's partial binding syntax:

```ls
@send _, _, response-to: id

File: src/websocket-connector/websocket-connector.ls [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+      id: uuid.v1!
+    request-data.payload = payload if payload?
+    request-data.response-to = options.response-to if options.response-to
+    @socket.send JSON.stringify request-data
+    @last-sent-id = request-data.id
+
+
+  listen: ->
+    @socket = new WebSocket "ws://#{@exocom-host}:#{@exocom-port}/services"
+      ..on 'open', ~>
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Extract an `on-socket-open` method

File: src/websocket-connector/websocket-connector.ls [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+        @emit 'online'
+      ..on 'message', (message) ~>
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Rename to `on-socket-message` and simplify to:

```ls
..on 'message', @_on-socket-message

File: src/websocket-connector/websocket-connector.ls [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+        @_on-message message
+      ..on 'error' (error) ~>
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Extract an `_on-socket-error` method

File: src/websocket-connector/websocket-connector.ls [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+
+
+  _log-sending: (message-name, options) ->
+    | options.response-to  =>  debug "sending message '#{message-name}' in response to '#{options.response-to}'"
+    | _                    =>  debug "sending message '#{message-name}'"
+
+
+  # Returns the relevant data from a request
+  _parse-request: (req) ->
+    message-name = req.name
```
Comment (kevgo) [inline src/websocket-connector/websocket-connector.ls]: Please change this in a separate PR to:

```ls
_parse-request: (req) ->
  {
    message-name: req.name
    payload: req.payload
    response-to: req.response-to
    id: req.id
  }

File: features/steps/port-steps.ls (hunk 10 above)
Comment (trushton) [inline features/steps/port-steps.ls]: In the previous `ExoComMock` PR, we made the ExoCom instance only save the messages that are not instances of `exocom.register-service`, therefore we can't do `wait-until (~> exocom.received.messages.length)`. Looking into the `service-sockets` object was my alternative solution, but you're right, just from looking at the `port-steps` there is no way to know that the service is `test-service`.

Comment (trushton) [inline features/steps/port-steps.ls]: Nevermind, this issue can be resolved by merging the `listen.feature` and `register-service.feature` as you recommend below.
